### ISSUE TITLE
Infer typeof self if unspecified on cast

### ISSUE THREAD

**Issue by alita-moore** [microsoft/TypeScript#45618]
# Suggestion

## 🔍 Search Terms

- type casting
- implicit typing
- syntactic sugar

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

It would be a lot cleaner if when you cast a type to some other type (for example NonNullable) you don't have to specify typeof self.

## 📃 Motivating Example

if you have an Array that could be undefined you can filter out undefined like so

```ts
const array = ["one", undefined, "two"].filter(Boolean);
console.log(array) // ["one", "two"]
```

but in this case the type would still be `(string | undefined)[]` so it would cause linting errors later in the code. While this may be specific to the .filter method, it may be unreasonable to expect all methods accurately account for type checks like this. So one easy way to fix this would just be to do

```ts
const arrayUndefined = ["one", undefined, "two"]

const array = arrayUndefined.filter(Boolean) as NonNullable<UnArrayify<typeof arrayUndefined>>
```

what I'm suggesting is that instead of requiring you do `typeof arrayUndefined` you can instead do an implicit cast to make the code cleaner
```ts
const arrayUndefined = ["one", undefined, "two"]

// it's implied that UnArrayify would be called with <typeof array.filter(Boolean)>
const array = arrayUndefined.filter(Boolean) as NonNullable<UnArrayify>
```

## 💻 Use Cases

I would like this because it would result in cleaner typing for unhandled edgecases in the typescript linter

**Comment by MartinJohns** [comment]
> would result in cleaner typing

The clean approach would be to use a type guard and avoid the type assertion (aka "cast") entirely.

Perhaps you can provide a better example?

Also you checked that this wouldn't be a breaking change, but introducing a new global type could very well be breaking for anyone having such a type defined.

**Comment by alita-moore** [comment]
@MartinJohns what do you mean by global type? I guess I don't know how this would be implemented from the Typescript side. But I don't know how it would involve a global type.

What do you mean by type guard? Do you mind providing an example

**Comment by alita-moore** [comment]
here's a couple examples that may help explain what I mean

```ts
type CastTo<T> = T

const array = ["string"]

// currently..
const str = array[0] as CastTo<number>;
str -> number
const str = array[0] as CastTo; // expected required type T

// I propose
const str = array[0] as CastTo<number>;
str -> number
const str = array[0] as CastTo; // i.e. CastTo is seen as CastTo<typeof array[0]> to the linter
str -> string
```

**Comment by alita-moore** [comment]
also it's not always possible to define type as non-nullable without running a potentially null array through a filter to assure the contents are defined.

I think the main reason this may cause issues right now though is because people might not notice that the type their casting to is being called with the implicit type which would be confusing. So maybe this isn't a great idea 🤷‍♀️

**Comment by MartinJohns** [comment]
> What do you mean by type guard? Do you mind providing an example

    function isNotUndefined<T>(value: T | undefined): value is T {
      return value !== undefined;
    }

    const array1 = ["one", undefined, "two"].filter(Boolean); // type is (string | undefined)[]
    const array2 = ["one", undefined, "two"].filter(isNotUndefined); // type is string[]

**Comment by jcalz** [comment]
As proposed, this would be a breaking change for any generic type with a [type parameter default](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html#generic-parameter-defaults).  For example, given `type Foo<T = string> = ...`, writing `123 as Foo` is currently the same as writing `123 as Foo<string>`, but you are suggesting it should be interpreted as `123 as Foo<123>` instead?

**Comment by alita-moore** [comment]
@jcalz it would only apply to required types, Foo<T = string> is an optional type. So it wouldn't affect those as proposed.

**Comment by alita-moore** [comment]
Also I encountered another use case where I want to make a type mutable (using type-fest) but I have to first define the const and then redfine it to get that type.

My exact example is

```ts
const _context = (await import("github")).context
const context = _context as Mutable<context>
```

So my proposed change would instead be this

```ts
const context = (await import("github")).context as Mutable
```

**Comment by MartinJohns** [comment]
You can achieve something like that trivially using a helper function without weird and confusing type shenanigans:

    function asMutable<T>(value: T): Mutable<T> { return value; }

**Comment by alita-moore** [comment]
> You can achieve something like that trivially using a helper function without weird and confusing type shenanigans:
> 
>     function asMutable<T>(value: T): Mutable<T> { return value; }
>     
>     const context = asMutable(await import("github")).context);

Yeah but it's still adding unnecessary complexity to my code. Which defeats the purpose

**Comment by alita-moore** [comment]
Checking in on this, I still think this would be a helpful addition.
